src/rules/preferReadonlyRule.ts

```diff
@@ -0,0 +1,158 @@
+ * limitations under the License.
+ */
+
+import * as utils from "tsutils";
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    public static metadata: Lint.IRuleMetadata = {
+        description: "Requires that private variables are marked as `readonly` if they're never modified outside of the constructor.",
+        descriptionDetails: Lint.Utils.dedent`
+            If a variable is only assigned to in the constructor, it should be declared as \`readonly\`.
+        `,
+        optionExamples: [true],
+        options: null,
+        optionsDescription: "Not configurable.",
+        ruleName: "prefer-readonly",
+        type: "maintainability",
+        typescriptOnly: true,
+    };
+
+    public static FAILURE_STRING_FACTORY(identifier: string) {
+        return `Private member variable '${identifier}' is never reassigned; mark it as 'readonly'.`;
+    }
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        const preferReadonlyWalker = new PreferReadonlyWalker(sourceFile, this.ruleName, undefined);
+        return this.applyWithWalker(preferReadonlyWalker);
+    }
+}
+
+class ClassScope {
+    public declaredMembers = new Map<string, ts.Identifier>();
+    public modifiedMembers = new Set<string>();
+    public inConstructor = false;
+}
+
+class PreferReadonlyWalker extends Lint.AbstractWalker<void> {
+    private scope: ClassScope;
+
+    public walk(sourceFile: ts.SourceFile) {
+        if (sourceFile.isDeclarationFile) {
+            return;
+        }
+
+        ts.forEachChild(sourceFile, this.visitNode);
+    }
+
+    private readonly visitNode = (node: ts.Node) => {
+        switch (node.kind) {
+            case ts.SyntaxKind.ClassDeclaration:
+                this.handleClassDeclaration(node as ts.ClassDeclaration);
+                break;
+
+            case ts.SyntaxKind.Constructor:
+                this.handleConstructor(node as ts.ConstructorDeclaration);
+                break;
+
+            case ts.SyntaxKind.PropertyDeclaration:
+                this.handlePropertyDeclaration(node as ts.PropertyDeclaration);
+                break;
+
+            case ts.SyntaxKind.PropertyAccessExpression:
+                this.handlePropertyAccessExpression(node as ts.PropertyAccessExpression, node.parent);
+                break;
+
+            default:
+                ts.forEachChild(node, this.visitNode);
+        }
+    }
+
+    private handleClassDeclaration(node: ts.ClassDeclaration) {
+        const parentScope = this.scope;
+        const childScope = this.scope = new ClassScope();
+
+        ts.forEachChild(node, this.visitNode);
+
+        this.finalizeScope(childScope);
+        this.scope = parentScope;
+    }
+
+    private handleConstructor(node: ts.ConstructorDeclaration) {
+        this.scope.inConstructor = true;
+        ts.forEachChild(node, this.visitNode);
+        this.scope.inConstructor = false;
```

Comment:
that's not sufficient, functions inside the constructor cannot assign readonly properties